File: tau.py
"""
Tau module: tau instead of radians, to make working in periods easier
"""

from cmath import rect as rad_rect, polar as rad_polar, pi
radian_cycle = pi * 2


def rad2tau(mag, rad):
    return mag, ((rad / radian_cycle) + 0.5) % 1 - 0.5


def tau2rad(mag, tau):
    return mag, tau * radian_cycle


def rect(period, mag=1):
    return rad_rect(*tau2rad(mag, period))


def polar(cval):
    return rad2tau(*rad_polar(cval))

File: tau.js
// JavaScript port of tau.py - works in tau units instead of radians.
export const radianCycle = Math.PI * 2;

function rad2tau(mag, rad) {
  // now exactly matches Python/C
  const x = rad/(2*Math.PI) + 0.5;
  const m = ((x % 1) + 1) % 1;       // true modulo into [0…1)
  return [ mag, m - 0.5 ];
}

export function tau2rad(mag, tau) {
  return [mag, tau * radianCycle];
}

export function rect(period, mag = 1) {
  const [m, rad] = tau2rad(mag, period);
  return { re: Math.cos(rad) * m, im: Math.sin(rad) * m };
}

export function polar(cval) {
  const mag = Math.hypot(cval.re, cval.im);
  const rad = Math.atan2(cval.im, cval.re);
  const [, tau] = rad2tau(mag, rad);
  return [mag, tau];
}

